internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,333 @@
+	if selector.MatchResources.ControlPlane != nil && class.Spec.ControlPlane.Ref != nil {
+		if selectorMatchTemplate(selector, class.Spec.ControlPlane.Ref) {
+			return nil
+		}
+		if class.Spec.ControlPlane.MachineInfrastructure != nil {
+			if selectorMatchTemplate(selector, class.Spec.ControlPlane.MachineInfrastructure.Ref) {
+				return nil
+			}
+		}
+	}
+	if selector.MatchResources.MachineDeploymentClass != nil {
+		for _, name := range selector.MatchResources.MachineDeploymentClass.Names {
+			for _, md := range class.Spec.Workers.MachineDeployments {
+				if md.Class == name {
+					if selectorMatchTemplate(selector, md.Template.Infrastructure.Ref) {
+						return nil
+					}
+					if selectorMatchTemplate(selector, md.Template.Bootstrap.Ref) {
+						return nil
+					}
+				}
+			}
+		}
+	}
+	// if the code has not returned at this point there is no matching template in the ClusterClass. Return an error.
+	return append(allErrs,
+		field.Invalid(
+			path,
+			selector,
+			"selector did not match any template in the ClusterClass",
+		))
+}
+
+// selectorMatchTemplate returns true if APIVersion and Kind for the given selector match the reference.
+func selectorMatchTemplate(selector clusterv1.PatchSelector, reference *corev1.ObjectReference) bool {
+	return selector.Kind == reference.Kind && selector.APIVersion == reference.APIVersion
+}
+
+var validOps = sets.NewString("add", "replace", "remove")
+
+func validateJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	variableSet := getClusterClassVariablesMap(variables)
+
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.NotSupported(
+					path.Index(i).Child("op"),
+					jsonPatch,
+					validOps.List(),
+				))
+		}
+
+		if !strings.HasPrefix(jsonPatch.Path, "/spec/") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of value or valueFrom.",
+				))
+		}
+
+		if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch can not define both value and valueFrom.",
+				))
+		}
+
+		// Attempt to marshal the JSON to discover if it  is valid.
+		if jsonPatch.Value != nil {
+			if err := json.Unmarshal(jsonPatch.Value.Raw, &struct{}{}); err != nil {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i),
+						jsonPatch.Value,
```

Comment:
```suggestion
						string(jsonPatch.Value.Raw),
```
otherwise it will just print an byte array which is not pretty

Reply:
I can use the .String()  of the Value - does that work for you?

Reply:
I would prefer the suggested variant. Example:
```
	value := apiextensionsv1.JSON{Raw: []byte(`{"cluster":{"name":"cluster-name","namespace":"default","topology":{"class":"clusterClass1","version":"v1.21.1"}}}`)}
	fmt.Println(value.String())
	fmt.Println(string(value.Raw))
```

Leads to:

```bash
&JSON{Raw:*[123 34 99 108 117 115 116 101 114 34 58 123 34 110 97 109 101 34 58 34 99 108 117 115 116 101 114 45 110 97 109 101 34 44 34 110 97 109 101 115 112 97 99 101 34 58 34 100 101 102 97 117 108 116 34 44 34 116 111 112 111 108 111 103 121 34 58 123 34 99 108 97 115 115 34 58 34 99 108 117 115 116 101 114 67 108 97 115 115 49 34 44 34 118 101 114 115 105 111 110 34 58 34 118 49 46 50 49 46 49 34 125 125 125],}

{"cluster":{"name":"cluster-name","namespace":"default","topology":{"class":"clusterClass1","version":"v1.21.1"}}}
```

Maybe I'm reproducing incorrectly, but that result doesn't look great :)

Reply:
Yikes - using the string()